src/algorithms/piecewise_linear_stretching.jl

```diff
@@ -0,0 +1,179 @@
+        for src_pair in src_intervals
+            if first(src_pair) == last(src_pair) 
+                throw(ArgumentError("The start and end of an interval in src_intervals cannot be equal."))
+            end
+        end
+        T = promote_type(T₁, T₂)
+        new{T}(src_intervals, dst_intervals)
+    end    
+end
+
+function (f::PiecewiseLinearStretching)(out::GenericGrayImage, img::GenericGrayImage) 
+    # Verify that the specified dst_intervals fall inside the permissible range
+    # of the output type. 
+    T = eltype(out)
+    for (a,b) in f.dst_intervals
+        p = typemin(T)
+        r = typemax(T)
+        if  !(p <= a <= r) || !(p <= b <= r)
+            throw(DomainError("The specified interval [$a, $b] falls outside the permissible range [$p, $r] associated with $T"))
+        end
+    end
+    # A linear mapping of a value x in the interval [A, B] to the interval [a,b] is given by
+    # the formula f(x) = (x-A) * ((b-a)/(B-A)) + a. The operation r * x - o is equivalent to
+    # (x-A) * ((b-a)/(B-A)) + a. Precalculate the r and o so that later on an inner loop
+    # contains only multiplication and addition to get better performance.
+    N = length(f.src_intervals)
+    rs = zeros(Float64, N)
+    os = zeros(Float64, N)
+    for (src, dst) in zip(pairs(f.src_intervals), pairs(f.dst_intervals)) 
+        n, (A,B) = src
+        _, (a,b) = dst   
+        rs[n] = (b - a) / (B - A)
+        os[n] = (A*b - B*a) / (B - A)
+    end
+
+    # Determine which interval the source pixel fall into, and apply the accompanying linear
+    # transformation taking into account NaN values. 
+    @inbounds for p in eachindex(img)
+        val = img[p]
+        if isnan(val)
+            out[p] = val
+        else
+            is_inside_src_intervals = false
+            for (n, (A,B)) in pairs(f.src_intervals)                                 
+                if (A <= val < B) || (A <= val <= B && n == N)       
+                    newval = rs[n] * val - os[n]                                                                          
+                    out[p] = T <: Integer ? round(Int, newval) : newval  
+                    is_inside_src_intervals = true  
+                    break                                                                                   
+                end                                                                                                    
+            end 
+            if !is_inside_src_intervals
+                # Saturate pixels that fall outside the specified src_intervals by setting
+                # them equal to the start/end of the edge dst_intervals. 
+                a = first(first(f.dst_intervals))
+                b = last(last(f.dst_intervals))
+                A = first(first(f.src_intervals))
+                newval = val < A ? a : b
+                out[p] = T <: Integer ? round(Int, newval) : newval  
+            end 

```

Comment:
What if

```julia
f = PiecewiseLinearStretching(src_intervals = (0.0 => 0.2, 0.8 => 1.0),
                              dst_intervals = (0.0 => 0.4, 0.9 => 1.0))
```

My understanding of this if-branch is that `0.5` will be mapped to `1.0`, which seems to be incorrect.

It seems that we need to ensure `src_intervals` is a cover of range `min(first(f.dst_intervals))..max(last(f.dst_intervals))`

Reply:
Excellent catch Johnny, thank you for that! This is indeed a serious oversight!